### PR TITLE
feat(route53): latency based routing

### LINKED ISSUE #28722: route53: support latency based routing
### Describe the feature

In Route 53, there is a latency-based routing feature. I want to implement this using CDK.

https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/routing-policy-latency.html

### Use Case

In Cfn, this can be implemented by setting the region argument, but it cannot be implemented from L2.

### Proposed Solution

Add region props to RecordSet and generate setIdentifier.
```ts
new RecordSet(scope, id, {
  ...  
  region: 'ap-northeast-1', // add
})
```

### Other Information

_No response_

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.116.1

### Environment details (OS name and version, etc.)

irrelevant

### PR DESCRIPTION
**Pull request by badmintoncryer**
In this PR, I have made the RecordSet compatible with latency-based routing.
You can set it by providing a Region name to the region argument.

```ts
new route53.ARecord(this, 'ARecord', {
  zone: myZone,
  target: route53.RecordTarget.fromIpAddresses('1.2.3.4'),
  region: 'ap-northeast-1', // added
})
```

Closes #28722.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-route53/README.md [hunk 1]
```diff
@@ -172,10 +184,10 @@ new route53.ARecord(this, 'ARecordWeighted1', {
   zone: myZone,
   target: route53.RecordTarget.fromIpAddresses('1.2.3.4'),
   weight: 10,
-  setIdentifier: 'weighted-record-id', 
+  setIdentifier: 'weighted-record-id',
 });
 ```
-**Warning** It is not possible to specify `setIdentifier` in a simple routing without one of `weight` or `geoLocation` defined.
+**Warning** It is not possible to specify `setIdentifier` in a simple routing without one of `weight`, `region` or `geoLocation` defined.
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-route53/README.md:190]:
```suggestion
**Warning** It is not possible to specify `setIdentifier` for a simple routing policy.
```

File: packages/aws-cdk-lib/aws-route53/lib/record-set.ts [hunk 2]
```diff
@@ -278,15 +292,24 @@ export class RecordSet extends Resource implements IRecordSet {
     if (props.setIdentifier && (props.setIdentifier.length < 1 || props.setIdentifier.length > 128)) {
       throw new Error(`setIdentifier must be between 1 and 128 characters long, got: ${props.setIdentifier.length}`);
     }
-    if (props.weight && props.geoLocation) {
-      throw new Error('Only one of weight or geoLocation can be specified, not both');
+    if (props.setIdentifier && !props.weight && !props.geoLocation && !props.region) {
+      throw new Error('setIdentifier can only be specified when either weight, geoLocation or region is specified');
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-route53/lib/record-set.ts:296]:
```suggestion
      throw new Error('setIdentifier can only be specified for non-simple routing policies');
```

File: packages/aws-cdk-lib/aws-route53/lib/record-set.ts [hunk 3]
```diff
@@ -200,6 +200,19 @@ export interface RecordSetOptions {
...
 
+  /**
+   * The Amazon EC2 Region where you created the resource that this resource record set refers to.
+   * The resource typically is an AWS resource, such as an EC2 instance or an ELB load balancer,
+   * and is referred to by an IP address or a DNS domain name, depending on the record type.
+   *
+   * When Amazon Route 53 receives a DNS query for a domain name and type for which you have created latency resource record sets,
+   * Route 53 selects the latency resource record set that has the lowest latency between the end user and the associated Amazon EC2 Region.
+   * Route 53 then returns the value that is associated with the selected resource record set.
+   *
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-route53/lib/record-set.ts:211]:
```suggestion
   *
   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-route53-recordset.html#cfn-route53-recordset-region
   *
```
Let's add the link to the documentation as there is additional information about the attribute constraints.

File: packages/aws-cdk-lib/aws-route53/lib/record-set.ts [hunk 4]
```diff
@@ -278,15 +292,24 @@ export class RecordSet extends Resource implements IRecordSet {
...
+    const propsToCheck = ['region', 'weight', 'geoLocation'];
+    let definedPropsCount = 0;
+    propsToCheck.forEach(prop => {
+      if (props[prop as keyof RecordSetProps] !== undefined) {
+        definedPropsCount++;
+      }
+    });
+    if (definedPropsCount > 1) {
+      throw new Error('Only one of region, weight, or geoLocation can be defined');
     }
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-route53/lib/record-set.ts:308]:
```suggestion
    let nonSimpleRoutingPolicies = 0;
    if (props.geoLocation) nonSimpleRoutingPolicies++;
    if (props.region) nonSimpleRoutingPolicies++;
    if (props.weight) nonSimpleRoutingPolicies++;
    if (nonSimpleRoutingPolicies > 1) {
      throw new Error('Only one of region, weight, or geoLocation can be defined');
    }
```
More concise.

File: packages/aws-cdk-lib/aws-route53/lib/record-set.ts [hunk 5]
```diff
@@ -369,12 +393,21 @@ export class RecordSet extends Resource implements IRecordSet {
...
+      const idPrefix= `REGION_${this.region}_ID_`;
+      return this.createIdentifier(idPrefix);
     }
 
     return undefined;
   }
+
+  private createIdentifier(prefix: string): string {
+    const identifier = `${prefix}${Names.uniqueResourceName(this, { maxLength: 64 - prefix.length })}`;
+    return identifier;
```
Comment (lpizzinidev) [inline packages/aws-cdk-lib/aws-route53/lib/record-set.ts:409]:
```suggestion
    return `${prefix}${Names.uniqueResourceName(this, { maxLength: 64 - prefix.length })}`;
```
The function name already clarifies what we are returning.

**Review by lpizzinidev — CHANGES_REQUESTED** [review]
Looks good overall 👍
I left some notes for minor adjustments.

**Comment by badmintoncryer** [comment]
@lpizzinidev Thank you for your review! I addressed your comments.

**Review by lpizzinidev — APPROVED** [review]
Nice 👍

File: packages/aws-cdk-lib/aws-route53/lib/record-set.ts [hunk 6]
```diff
@@ -278,15 +294,21 @@ export class RecordSet extends Resource implements IRecordSet {
...
+    if (props.setIdentifier && !props.weight && !props.geoLocation && !props.region) {
+      throw new Error('setIdentifier can only be specified for non-simple routing policies');
     }
-    if (props.setIdentifier && !props.weight && !props.geoLocation) {
-      throw new Error('setIdentifier can only be specified when either weight or geoLocation is defined');
+
+    let nonSimpleRoutingPolicies = 0;
+    if (props.geoLocation) nonSimpleRoutingPolicies++;
+    if (props.region) nonSimpleRoutingPolicies++;
+    if (props.weight) nonSimpleRoutingPolicies++;
```
Comment (aaythapa) [inline packages/aws-cdk-lib/aws-route53/lib/record-set.ts:304]:
```suggestion
    let nonSimpleRoutingPolicies = [props.geoLocation, props.region, props.weight].filter((variable) => variable !== undefined).length;
```

**Review by aaythapa — COMMENTED** [review]
Thank you for your contribution! Looks great overall, only concern is should the region names be enum values? The docs clearly define which regions are allowed in this property so maybe we should do the same

**Comment by badmintoncryer** [comment]
@aaythapa Thank you for your review. I'll add enum later!

**Comment by badmintoncryer** [comment]
@aaythapa I've added LatencyBasedRoutingRegion. Please comfirm it.

**Comment by aaythapa** [comment]
> @aaythapa I've added LatencyBasedRoutingRegion. Please confirm it.

Hi @badmintoncryer thank you for making the changes. I talked it over with the team and we decided we don't want to add enums for regions. Our reasoning is users already use strings when defining region in CDK, for example [here](https://github.com/aws/aws-cdk/blob/d46b5359cf0c4960b6aeae989b72278df7e49a4e/packages/aws-cdk-lib/aws-dynamodb/lib/table-v2.ts#L133) and with references like `stack.env.region`. On top of that, new regions are constantly being added so we need to manually maintain the list of enums which add additional overhead. 

Sorry about the extra work, I should've discussed this with the team before suggesting you add enums. Hopefully it isn't too much work to go back.

**Comment by badmintoncryer** [comment]
@aaythapa I've reverted modification! Please confirm it again.

**Pull request merged**